### PR TITLE
Sodium hash

### PR DESCRIPTION
**Pull request by phadej**
Implementing `Hash` module using `libsodium`.

### CODE REVIEW

**Comment by phadej** [comment]
@dcoutts this is ready for review

File: cardano-crypto-class/src/Cardano/Crypto/Libsodium/Constants.hsc [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    CRYPTO_SHA512_STATE_SIZE,
+    CRYPTO_BLAKE2B_256_STATE_SIZE,
+    )  where
+
+#include <sodium.h>
+
+-- From https://libsodium.gitbook.io/doc/advanced/sha-2_hash_function
+-- and https://libsodium.gitbook.io/doc/hashing/generic_hashing
+
+type CRYPTO_SHA256_BYTES = #{const crypto_hash_sha256_BYTES}
```
Comment (nc6) [inline cardano-crypto-class/src/Cardano/Crypto/Libsodium/Constants.hsc:16]: Could use pattern synonyms for these, and then we can match on them

File: cardano-crypto-class/src/Cardano/Crypto/Libsodium/Hash.hs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+digestMLockedFB = digestMLockedStorable
+
+digestMLockedBS
+    :: forall h proxy. (SodiumHashAlgorithm h)
+    => proxy h -> SB.ByteString -> IO (MLockedFiniteBytes (SizeHash h))
+digestMLockedBS p bs = SB.useAsCStringLen bs $ \(ptr, len) -> do
+    digestMLocked p (castPtr ptr) len
+
+-------------------------------------------------------------------------------
+-- Hash expansion
```
Comment (nc6) [inline cardano-crypto-class/src/Cardano/Crypto/Libsodium/Hash.hs:73]: Can you provide some more detail here? What is hash expansion, and why would I want to do it?

Comment (nc6) [reply cardano-crypto-class/src/Cardano/Crypto/Libsodium/Hash.hs:73]: Oh, or this is specific to the SUM KES construction?

File: cardano-crypto-class/src/Cardano/Crypto/Libsodium/Memory/Internal.hs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+finalizeMLockedForeignPtr :: forall a. MLockedForeignPtr a -> IO ()
+finalizeMLockedForeignPtr = coerce (finalizeForeignPtr @a)
+
+traceMLockedForeignPtr :: (Storable a, Show a) => MLockedForeignPtr a -> IO ()
+traceMLockedForeignPtr fptr = withMLockedForeignPtr fptr $ \ptr -> do
+    a <- peek ptr
+    print a
+
+{-# DEPRECATED traceMLockedForeignPtr "Don't leave traceMLockedForeignPtr in production" #-}
```
Comment (nc6) [inline cardano-crypto-class/src/Cardano/Crypto/Libsodium/Memory/Internal.hs:44]: Ah, deprecation is for this? Makes sense..

File: cardano-crypto-class/memory-example/Main.hs [hunk 4]
```diff
@@ -12,7 +16,11 @@ import           System.Environment (getArgs)
 import           System.Posix.Process (getProcessID)
 #endif
 
-import           Cardano.Crypto.Libsodium (allocSecureForeignPtr, sodiumInit)
+import qualified Data.ByteString as SB
```
[the remark this answers is not in the document]
Comment (phadej) [reply cardano-crypto-class/memory-example/Main.hs:19]: This was copied from `Cardano.Crypto.Hash.Class`

```
cardano-crypto-class/memory-example/Main.hs:import qualified Data.ByteString as SB
cardano-crypto-class/src/Cardano/Crypto/Hash/Class.hs:import qualified Data.ByteString as SB
cardano-crypto-class/src/Cardano/Crypto/Hash/Class.hs:import qualified Data.ByteString.Char8 as SB8
cardano-crypto-class/src/Cardano/Crypto/Libsodium/Hash.hs:import qualified Data.ByteString as SB
cardano-crypto-tests/src/Test/Crypto/Hash.hs:import qualified Data.ByteString as SB
```

should I change the rest too?

---

FWIW: HLint has an option to say how qualified modules should be named. E.g. https://github.com/phadej/phadej-universe/blob/2f1cb7169b89802fd9180e09d2f4397a06917d5b/z-scripts/.hlint.yaml#L69-L90

File: cardano-crypto-class/src/Cardano/Crypto/FiniteBytes.hs [hunk 5]
```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE MagicHash           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UnboxedTuples       #-}
+module Cardano.Crypto.FiniteBytes
```
[the remark this answers is not in the document]
Comment (phadej) [reply cardano-crypto-class/src/Cardano/Crypto/FiniteBytes.hs:6]: I figured out that I should use `PinnedFiniteBytes`, and maybe rename it to `PinnedSizedBytes` which would be a bit more obvious name, but will add the documentation there.

File: cardano-crypto-class/src/Cardano/Crypto/Libsodium/Constants.hsc (hunk 1 above)
Comment (phadej) [reply cardano-crypto-class/src/Cardano/Crypto/Libsodium/Constants.hsc:16]: Note: `type`, not value.

File: cardano-crypto-class/src/Cardano/Crypto/Libsodium/Hash.hs (hunk 2 above)
Comment (phadej) [reply cardano-crypto-class/src/Cardano/Crypto/Libsodium/Hash.hs:73]: Yes this is specific to  SUM KES. The current name for this is `splitSeed`, but that one isn't good either as it uses *some* hashing (and produces seed of size specific to the hash function).

Better name would help.

FWIW, sodium's blake2b implementation allows to pass *key* to the hash function, thus AFAICS salting the result, which we do here manually. See https://libsodium.gitbook.io/doc/hashing/generic_hashing#usage

File: cardano-crypto-class/src/Cardano/Crypto/Libsodium/Memory/Internal.hs (hunk 3 above)
Comment (phadej) [reply cardano-crypto-class/src/Cardano/Crypto/Libsodium/Memory/Internal.hs:44]: Yes, makes `-Werror` useful and not only annoying ;)

File: cardano-crypto-class/memory-example/Main.hs [hunk 6]
```diff
@@ -1,9 +1,13 @@
 {-# LANGUAGE CPP #-}
 {-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -Wno-deprecations #-}
```
[the remark this answers is not in the document]
Comment (phadej) [reply cardano-crypto-class/memory-example/Main.hs:7]: use of traceMLockedFiniteBytes, which shouldn't be present in non-testing/debug mode. Pushed a commit adding a comment.

File: cardano-crypto-class/src/Cardano/Crypto/Libsodium/Constants.hsc (hunk 1 above)
Comment (nc6) [reply cardano-crypto-class/src/Cardano/Crypto/Libsodium/Constants.hsc:16]: Yep, sorry, saw how you were using it below.

File: cardano-crypto-class/memory-example/Main.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (phadej) [reply cardano-crypto-class/memory-example/Main.hs]: Renamed

File: cardano-crypto-class/src/Cardano/Crypto/FiniteBytes.hs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        res <- c_sodium_compare_unsafe (byteArrayContents x) (byteArrayContents y) size
+        return (compare res 0)
+      where
+        size :: CSize
+        size = fromInteger (natVal (Proxy :: Proxy n))
+
+-- |
+--
+-- If given 'String' is too long, it is truncated,
+-- If it is too short, it is padded with zeros.
```
[the remark this answers is not in the document]
Comment (phadej) [reply cardano-crypto-class/src/Cardano/Crypto/FiniteBytes.hs]: Done

**Comment by phadej** [comment]
part of #138

**Pull request closed without merging**